packages/gatsby/src/utils/source-nodes.ts

```diff
@@ -74,23 +74,71 @@ function getStaleNodes(
       )
     }
 
+    if (state.statefulSourcePlugins.has(rootNode.internal.owner)) {
+      return false
+    }
+
     return !state.nodesTouched.has(rootNode.id)
   })
 }
 
 /**
- * Find all stale nodes and delete them
+ * Find all stale nodes and delete them unless the node type has been opted out of stale node garbage collection.
  */
-function deleteStaleNodes(
-  state: IGatsbyState,
-  nodes: GatsbyIterable<IGatsbyNode>
-): void {
-  const staleNodes = getStaleNodes(state, nodes)
+async function deleteStaleNodes(
+  previouslyExistingNodeTypeNames: Array<string>
+): Promise<void> {
+  const state = store.getState()
+
+  let deleteCount = 0
+
+  const cleanupStaleNodesActivity =
+    report.createProgress(`Clean up stale nodes`)
+
+  cleanupStaleNodesActivity.start()
+
+  const { pluginNamesToOwnedNodeTypes, statefulSourcePlugins } = state
+
+  const typeNamesToOwnerPluginName = new Map()
+
+  pluginNamesToOwnedNodeTypes.forEach((ownedTypes, pluginName) => {
+    ownedTypes.forEach(typeName => {
+      typeNamesToOwnerPluginName.set(typeName, pluginName)
+    })
+  })
 
-  staleNodes.forEach(node => store.dispatch(deleteNode(node)))
+  for (const typeName of previouslyExistingNodeTypeNames) {
```

Comment:
`previouslyExistingNodeTypeNames` is passed in, now only types that previously existed are checked for staleness. Anything newly created types are not stale